File: prog3/trab1/src/Prog3_Trab1/CsvInput.java
package Prog3_Trab1;

import java.io.File;
import java.io.IOException;
import java.text.NumberFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.Scanner;
import java.util.concurrent.Callable;

final class CsvInput {
	
	private String CsvFilePath; 
	private String[] currentBufferedFields;
	private MediaCollection mediaCollection;
	
	public CsvInput(MediaCollection mediaCollection) {
		this.mediaCollection = mediaCollection;
	}
	
	private void readCsvInput(Callable<Void> specialReadMethod) throws ProgramException {
		
		try(Scanner CsvFileScanner = new Scanner(new File(this.CsvFilePath),"UTF-8")) {
			String line = CsvFileScanner.nextLine();
			while(CsvFileScanner.hasNextLine()) {
				line = CsvFileScanner.nextLine();
				this.currentBufferedFields = line.split(";");
				
				try { specialReadMethod.call(); }
				catch (DataFormatException dfe) {
					System.out.println("Erro de formatação");
					throw new ProgramException();
				}
				catch (InconsistentDataException ide) {
					System.out.println("Dados inconsistentes " + ide.getDataLocation());
					throw new ProgramException();
				}
				catch(Exception e) {
					System.out.println("Unexpected Exception");
					throw new ProgramException();
				}
			}
			this.currentBufferedFields = null;
		}
		catch (IOException e) {
			System.out.println("Erro de I/O");
			throw new ProgramException();
		}
	}
	
	private void insertGenreFromString() throws DataFormatException {
		String[] lineFields = this.currentBufferedFields;
		if(lineFields.length == 2) {
			String genreInitials = lineFields[0];
			String genreName = lineFields[1];
			Genre genre = new Genre(genreInitials);
			genre.setName(genreName);
			this.mediaCollection.addGenre(genre);
		}
		else {
			throw new DataFormatException();
		}
	}
	
	private void insertPersonFromString() throws DataFormatException {
		String[] lineFields = this.currentBufferedFields;
		if(lineFields.length == 2) {
			long personId;
			try { personId = Long.parseLong(lineFields[0]); }
			catch(NumberFormatException e) {
				throw new DataFormatException();
			}
			String personName = lineFields[1];
			Person person = new Person(personId);
			person.setPersonName(personName);
			this.mediaCollection.addPerson(person);
		}
		else {
			throw new DataFormatException();
		}
	}
	
	private Media insertBookFromString() throws DataFormatException, InconsistentDataException {
		String[] lineFields = this.currentBufferedFields;
		long mediaId;
		int numberPages;
		try {
			mediaId = Long.parseLong(lineFields[0]);
			numberPages = Integer.parseInt(lineFields[5]);
		}
		catch(NumberFormatException e) {
			throw new DataFormatException();
		}
		String mediaName = lineFields[1];
		Genre genre = this.mediaCollection.getGenreByInitials(lineFields[6]);
		Book book = new Book(mediaId,mediaName,numberPages,genre);
		
		String[] authorCodesStr = lineFields[4].split(",");
		for(int i = 0; i < authorCodesStr.length; i++) {
			long authorCode;
			try { authorCode = Long.parseLong(authorCodesStr[i]); }
			catch(NumberFormatException e) {
				throw new DataFormatException();
			}
			Person person = this.mediaCollection.getPersonByCode(authorCode);
			if(person == null) {
				throw new InconsistentDataException("Autor", authorCodesStr[i]);
			}
			book.addAuthors(person);
		}
		return (Media) book;
	}
	
	private Media insertMovieFromString() throws DataFormatException, InconsistentDataException {
		String[] lineFields = this.currentBufferedFields;
		long mediaId, directorId;
		int featureLength;
		try {
			mediaId = Long.parseLong(lineFields[0]);
			featureLength = Integer.parseInt(lineFields[5]);
			directorId = Integer.parseInt(lineFields[3]);
		}
		catch(NumberFormatException e) {
			throw new DataFormatException();
		}
		String mediaName = lineFields[1];
		Genre genre = this.mediaCollection.getGenreByInitials(lineFields[6]);
		Person director = this.mediaCollection.getPersonByCode(directorId);
		if(director == null) {
			throw new InconsistentDataException("Diretor", Long.toString(directorId));
		}	
		Movie movie = new Movie(mediaId,mediaName,director,featureLength,genre);
		
		if(lineFields[4].length() == 0) return (Media) movie;
		
		String[] actorCodesStr = lineFields[4].split(",");
		for(int i = 0; i < actorCodesStr.length; i++) {
			long actorCode;
			try { actorCode = Long.parseLong(actorCodesStr[i]); }
			catch(NumberFormatException e) {
				throw new DataFormatException(); 
			}
			Person person = this.mediaCollection.getPersonByCode(actorCode);
			if(person == null) {
				throw new InconsistentDataException("Ator", actorCodesStr[i]);
			}
			movie.addMainActors(person);
		}
		return (Media) movie;
	}
	
	private Media insertTvSeriesSeasonFromString() throws DataFormatException, InconsistentDataException {
		String[] lineFields = this.currentBufferedFields;
		long mediaId;
		int featureLength, seasonNumber;
		try {
			mediaId = Long.parseLong(lineFields[0]);
			featureLength = Integer.parseInt(lineFields[5]);
			seasonNumber = Integer.parseInt(lineFields[8]);
		}
		catch(NumberFormatException e) {
			throw new DataFormatException();
		}
		String mediaName = lineFields[1];
		Genre genre = this.mediaCollection.getGenreByInitials(lineFields[6]);
		String seriesName = lineFields[7];
		
		TvSeriesSeason season = new TvSeriesSeason(mediaId, mediaName, seriesName,
				featureLength, seasonNumber, genre);
		
		if(lineFields[4].length() == 0) return (Media) season;
		
		String[] actorCodesStr = lineFields[4].split(",");
		for(int i = 0; i < actorCodesStr.length; i++) {
			long actorCode;
			try { actorCode = Long.parseLong(actorCodesStr[i]); }
			catch(NumberFormatException e) {
				throw new DataFormatException();
			}
			Person person = this.mediaCollection.getPersonByCode(actorCode);
			if(person == null) {
				throw new InconsistentDataException("Ator", actorCodesStr[i]);
			}
			season.addMainActors(person);
		}
		return (Media) season;
	}
	
	private void insertMediaFromString() throws DataFormatException, InconsistentDataException {
		String[] lineFields = this.currentBufferedFields;
		if(lineFields.length == 13) {
			Media media = null;
			
			String genreInitials = lineFields[6];
			Genre genre = this.mediaCollection.getGenreByInitials(genreInitials);
			if(genre == null) {
				throw new InconsistentDataException("Gênero", genreInitials);
			}
			
			if(lineFields[2].equals("L")) {
				media = this.insertBookFromString();
			}
			else if(lineFields[2].equals("F")) {
				media = this.insertMovieFromString();
			}
			else if(lineFields[2].equals("S")) {
				media = this.insertTvSeriesSeasonFromString();
			}
			else { throw new InconsistentDataException("Tipo", lineFields[2]); }
			
			try {
				double price = NumberFormat.getInstance(Locale.FRANCE).parse(lineFields[12]).doubleValue();
				media.setPrice(price);
			} catch (ParseException e) {
				throw new DataFormatException();
			}
			media.setConsumptionStatus(Media.WANT,lineFields[11].equals("x"));
			media.setConsumptionStatus(Media.WATCHED,lineFields[10].equals("x"));
			media.setConsumptionStatus(Media.NOTWATCHED,lineFields[9].equals("x"));
			this.mediaCollection.addMedia(media);
		}
		else {
			throw new DataFormatException();
		}
	}
	
	private void insertLoanFromString() throws DataFormatException, InconsistentDataException {
		String[] lineFields = this.currentBufferedFields;
		if(lineFields.length == 4) {
			int mediaId;
			try { mediaId = Integer.parseInt(lineFields[0]); }
			catch(NumberFormatException e) {
				throw new DataFormatException();
			}
			
			String lentTo = lineFields[1];
			SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");
			Date limitDate;
			Date startDate;
			try {
				startDate = dateFormat.parse(lineFields[2]);
				limitDate = dateFormat.parse(lineFields[3]);
			}
			catch (ParseException e) {
				throw new DataFormatException();
			}
			
			Media media = this.mediaCollection.getMediaById(mediaId);
			if(media == null) {
				throw new InconsistentDataException("Mídia", Integer.toString(mediaId));
			}
			Loan loan = new Loan(lentTo,media,startDate,limitDate);
			this.mediaCollection.lend(loan);
		}
		else {
			throw new DataFormatException();
		}
	}
	
	public void readGenreCsv(String filePath) throws ProgramException {
		this.CsvFilePath = filePath;
		Callable<Void> readInput = new Callable<Void>() {

			@Override
			public Void call() throws DataFormatException {
				insertGenreFromString();
				return null;
			}
		};
		this.readCsvInput(readInput);
	}
	
	public void readPersonCsv(String filePath) throws ProgramException {
		this.CsvFilePath = filePath;
		Callable<Void> readInput = new Callable<Void>() {

			@Override
			public Void call() throws DataFormatException {
				insertPersonFromString();
				return null;
			}
		};
		this.readCsvInput(readInput);
	}
	
	public void readMediaCsv(String filePath) throws ProgramException {
		this.CsvFilePath = filePath;
		Callable<Void> readInput = new Callable<Void>() {

			@Override
			public Void call() throws DataFormatException, InconsistentDataException {
				insertMediaFromString();
				return null;
			}
		};
		this.readCsvInput(readInput);
	}
	
	public void readLoanCsv(String filePath) throws ProgramException {
		this.CsvFilePath = filePath;
		Callable<Void> readInput = new Callable<Void>() {

			@Override
			public Void call() throws DataFormatException, InconsistentDataException {
				insertLoanFromString();
				return null;
			}
		};
		this.readCsvInput(readInput);
	}
	
}

File: prog3/trab2/CsvInput.cpp

#include "CsvInput.hpp"

template <class T>
inline std::string to_string (const T& t)
{
    std::stringstream ss;
    ss << t;
    return ss.str();
}

std::vector<std::string> split(std::string str,std::string sep){
    char* cstr=const_cast<char*>(str.c_str());
    char* current;
    std::vector<std::string> arr;
    current=strtok(cstr,sep.c_str());
    while(current!=NULL){
        arr.push_back(current);
        current=strtok(NULL,sep.c_str());
    }
    return arr;
}

string& ltrim(string &s) {
    s.erase(s.begin(), find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace))));
    return s;
}

string& rtrim(string &s) {
    s.erase(find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(), s.end());
    return s;
}

string& trim(string& s) {
    return ltrim(rtrim(s));
}

// Constructor
CsvInput::CsvInput(MediaCollection& mediaCollection) : mediaCollection(mediaCollection) {}


/*Private*/ void CsvInput::readCsvFile(void (CsvInput::*dataHandler) (const vector<std::string>&)) {

	string record;
	vector<string> fields;
	const char* filePath = this->csvFilePath.c_str();
	ifstream csvInputFile (filePath);

    try {
        if(!csvInputFile.good()) {
            throw IOException();
        }

        getline(csvInputFile,record);//Header line

        while(getline(csvInputFile,record)) {

            Tokenizer* lineTokenizer = new Tokenizer(record,';');

            while(lineTokenizer->hasNext()) {
                string line = lineTokenizer->next();
                line = trim(line);
                fields.push_back(line);
            }

            (this->*dataHandler)(fields);
            fields.clear();
            delete lineTokenizer;
        }

        csvInputFile.close();
    }
    catch(IOException e) {
        cout << e.what() << endl;
        throw ProgramException();
    }
    catch(DataFormatException e) {
        cout << e.what() << endl;
        throw ProgramException();
    }
    catch(InconsistentDataException e) {
        cout << e.what() << endl;
        throw ProgramException();
    }
}

/*Private*/ void CsvInput::insertGenreFromString(const vector<std::string>& fields) {

	string genreInitials;
	string genreName;

	if(fields.size() == 2) {
		genreInitials = fields[0];
		genreName = fields[1];
		Genre* newGenre = new Genre(genreInitials,genreName);
		this->mediaCollection.addGenre(*newGenre);
		if(this->mediaCollection.getGenreByInitials(genreInitials) != nullptr) {
		}
	}
	else {
		throw DataFormatException();
		//Error: Data Format Exception
	}
}

/*Private*/ void CsvInput::insertPersonFromString(const vector<std::string>& fields) {

	string personName;
	long personId;

	if(fields.size() == 2) {
        try {
            if(!isNumber(fields[0])) throw DataFormatException();
            personId = stol(fields[0],nullptr,10);/*Error: Handle DataFormatException*/
            personName = fields[1];
            Person* person = new Person(personId,personName);
            this->mediaCollection.addPerson(*person);
        }
        catch(invalid_argument& e) {
            throw DataFormatException();
        }
	}
	else {
		throw DataFormatException();
		//Error: Data Format Exception
	}
}

/*Private*/ Media* CsvInput::insertBookFromString(const vector<std::string>& fields) {

	unsigned int i;
	long mediaId,authorCode;
	int numberPages;
	Genre *genrePtr;
	string mediaName;


	mediaId = stol(fields[0],nullptr,10);
	numberPages = stoi(fields[5],nullptr,10);
	mediaName = fields[1];
	genrePtr = this->mediaCollection.getGenreByInitials(fields[6]);
	if(genrePtr == nullptr) {
        throw InconsistentDataException("Gênero", fields[6]);
		//Error: genre not found
	}

	Book* book = new Book(mediaId,mediaName,numberPages,*genrePtr);
	vector<string> authorCodesStr = split(fields[4],",");
	for(i=0;i < authorCodesStr.size() ;i++) {
        try {
            if(!isNumber(authorCodesStr[i])) throw DataFormatException();
            authorCode = stol(authorCodesStr[i],nullptr,10); //Error: Possible number conversion error.
            Person* person = (this->mediaCollection.getPersonByCode(authorCode));
            if(person == nullptr) {
                throw InconsistentDataException("Pessoa", to_string(authorCode));
            }
            book->addAuthor(*person);
        }
        catch(invalid_argument& e) {
            throw DataFormatException();
        }
	}
	return (Media*) book;
}

/*Private*/ Media* CsvInput::insertMovieFromString(const vector<std::string>& fields) {

	long mediaId, directorId,actorCode;
	int featureLength;
	unsigned int i;
	string mediaName;

	mediaId = stol(fields[0],nullptr,10);
	featureLength = stol(fields[5],nullptr,10);
    if(!isNumber(fields[3])) throw DataFormatException();
	directorId = stol(fields[3],nullptr,10);

	mediaName = fields[1];

	Genre* genrePtr = this->mediaCollection.getGenreByInitials(fields[6]);
	Person* directorPtr = this->mediaCollection.getPersonByCode(directorId);

	if(directorPtr == nullptr) {
        throw InconsistentDataException("Diretor", fields[3]);
		//Error: InconsistentDataException
	}

	if(genrePtr == nullptr) {
        throw InconsistentDataException("Gênero", fields[6]);
		//Error: InconsistentDataException
	}

	Movie* movie = new Movie(mediaId,mediaName,*directorPtr,featureLength,*genrePtr);
	vector<string> actorCodesStr = split(fields[4],",");
	for(i=0;i<actorCodesStr.size();i++) {
        try {
            if(!isNumber(actorCodesStr[i])) throw DataFormatException();
            actorCode = stol(actorCodesStr[i],nullptr,10); //Error: possible DataFormatException
            Person* actorPtr = this->mediaCollection.getPersonByCode(actorCode);
            if(actorPtr == nullptr) {
                throw InconsistentDataException("Ator", to_string(actorCode));
                //Error: possble InconsistentDataException
            }
            movie->addActor(*actorPtr);
        }
        catch(invalid_argument& e) {
            throw DataFormatException();
        }
	}
	return (Media*) movie;
}

/*Private*/ Media* CsvInput::insertTvSeriesSeasonFromString(const vector<std::string>& fields) {

	long mediaId,actorCode;
	unsigned int i;
	int featureLength, seasonNumber;
	string mediaName,seriesName;

    try {
        mediaId = stol(fields[0],nullptr,10); //Error possible DataFormatException
        featureLength = stol(fields[5],nullptr,10);
        if(!isNumber(fields[8])) throw DataFormatException();
        seasonNumber = stol(fields[8],nullptr,10);

        mediaName = fields[1];
        seriesName = fields[7];
        Genre* genrePtr = this->mediaCollection.getGenreByInitials(fields[6]);
        if(genrePtr == nullptr) {
            throw InconsistentDataException("Gênero", fields[6]);
            //Error: possible InconsistentDataException
        }

        TvSeriesSeason* season = new TvSeriesSeason(mediaId, mediaName, seriesName,featureLength, seasonNumber, *genrePtr);

        vector<string> actorCodesStr = split(fields[4],",");

        for(i=0;i<actorCodesStr.size();i++) {
            if(!isNumber(actorCodesStr[i])) throw DataFormatException();
            actorCode = stol(actorCodesStr[i],nullptr,10); //Error: possible DataFormatException
            Person* actorPtr = this->mediaCollection.getPersonByCode(actorCode);
            if(actorPtr == nullptr) {
                throw InconsistentDataException("Ator", to_string(actorCode));
                //Error: possible InconsistentDataException
            }
            season->addActor(*actorPtr);
        }

        return season;
    }
    catch(invalid_argument& e) {
        throw DataFormatException();
    }
}

/*Private*/ void CsvInput::insertMediaFromString(const vector<std::string>& fields) {

	string genreInitials,priceStr;
	double mediaPrice;
	Media* media;

	if(fields.size() == 13) {

        if(!isNumber(fields[0])) throw DataFormatException();
		genreInitials = fields[6];
		Genre* genrePtr = this->mediaCollection.getGenreByInitials(genreInitials);
		if(genrePtr == nullptr) {
            throw InconsistentDataException("Gênero", genreInitials);
			//Error: possible InconsistentDataException
		}

		if(fields[2] == "L") {
			media = this->insertBookFromString(fields);
		}
		else if(fields[2] == "F") {
			media = this->insertMovieFromString(fields);
		}
		else if(fields[2] == "S") {
			media = this->insertTvSeriesSeasonFromString(fields);
		}
		else {
            throw InconsistentDataException("Tipo", fields[2]);
			//Error: possible InconsistentDataException
		}

		priceStr = fields[12];
        for(string::iterator it = priceStr.begin(); it != priceStr.end(); it++) {
            if(*it == ',')
                *it = '.';
        }
		mediaPrice = stod(priceStr.c_str());
		media->setPrice(mediaPrice);

		media->setConsumptionStatus(Media::WANT,fields[11] == "x");

		media->setConsumptionStatus(Media::WATCHED,fields[10] == "x");

		media->setConsumptionStatus(Media::NOTWATCHED,fields[9] == "x");


		this->mediaCollection.addMedia(*media);
	}
	else {
		throw DataFormatException();
	}
}

/*Private*/ void CsvInput::insertLoanFromString(const vector<std::string>& fields) {

	long mediaId;
	string lentTo;
	time_t limitDate,startDate;
	Media* mediaPtr;

	if(fields.size() != 4) {
		throw DataFormatException();
		//Error: DataFormatException
	}

    try {
        if(!isNumber(fields[0])) throw DataFormatException();
        mediaId = stol(fields[0],nullptr,10);
        lentTo = fields[1];
    }
    catch(invalid_argument& e) {
        throw DataFormatException();
    }

    if(!validDate(fields[2], DATE_FORMAT) || !validDate(fields[3], DATE_FORMAT))
        throw DataFormatException();

    limitDate = parseDate(fields[2],DATE_FORMAT); // Error: possible DataFormatException
    startDate = parseDate(fields[3],DATE_FORMAT); // Error: possible DataFormatException

	mediaPtr = this->mediaCollection.getMediaById(mediaId);

	if(mediaPtr == nullptr) {
        throw InconsistentDataException("Midia", to_string(mediaId));
		//Error: possible InconsistentDataException
	}

	Media& MediaRef = *mediaPtr;

	Loan* loan = new Loan(lentTo,MediaRef,startDate,limitDate);
	this->mediaCollection.lend(*loan);

}

/*Public*/ void CsvInput::readGenreCsv(string filePath) {
	this->csvFilePath = filePath;
	this->readCsvFile(&CsvInput::insertGenreFromString);
}

/*Public*/ void CsvInput::readPersonCsv(string filePath) {
	this->csvFilePath = filePath;
	this->readCsvFile(&CsvInput::insertPersonFromString);
}

/*Public*/ void CsvInput::readMediaCsv(string filePath) {
	this->csvFilePath = filePath;
	this->readCsvFile(&CsvInput::insertMediaFromString);
}

/*Public*/ void CsvInput::readLoanCsv(string filePath) {
	this->csvFilePath = filePath;
	this->readCsvFile(&CsvInput::insertLoanFromString);
}

CsvInput::~CsvInput() {}
